cider/exec/operator/join/CiderLinearProbingHashtable.h

```diff
@@ -0,0 +1,273 @@
+    friend ContT;
+  };
+
+  using iterator = hm_iterator<LPHashTable, value_type>;
+  using const_iterator = hm_iterator<const LPHashTable, const value_type>;
+
+ public:
+  LPHashTable(size_type bucket_count,
+              Key empty_key,
+              const allocator_type& alloc = allocator_type())
+      : empty_key_({empty_key, false, 0}), buckets_(alloc) {
+    size_t pow2 = 1;
+    while (pow2 < bucket_count) {
+      pow2 <<= 1;
+    }
+    buckets_.resize(pow2, std::make_pair(empty_key_, T()));
+  }
+
+  LPHashTable(const LPHashTable& other, size_type bucket_count)
+      : LPHashTable(bucket_count, other.empty_key_.key, other.get_allocator()) {
+    for (auto it = other.begin(); it != other.end(); ++it) {
+      insert(*it);
+    }
+  }
+
+  allocator_type get_allocator() const noexcept { return buckets_.get_allocator(); }
+
+  // Iterators
+  iterator begin() noexcept { return iterator(this); }
+
+  const_iterator begin() const noexcept { return const_iterator(this); }
+
+  const_iterator cbegin() const noexcept { return const_iterator(this); }
+
+  iterator end() noexcept { return iterator(this, buckets_.size()); }
+
+  const_iterator end() const noexcept { return const_iterator(this, buckets_.size()); }
+
+  const_iterator cend() const noexcept { return const_iterator(this, buckets_.size()); }
+
+  // Capacity
+  bool empty() const noexcept { return size() == 0; }
+
+  size_type size() const noexcept { return size_; }
+
+  size_type max_size() const noexcept { return buckets_.max_size() / 2; }
+
+  bool insert(const std::pair<key_type, T>& value) {
+    return emplace_impl(value.first.key, value.second);
+  }
+
+  bool insert(std::pair<key_type, T>&& value) {
+    return emplace_impl(value.first.key, std::move(value.second));
+  }
+
+  template <typename... Args>
+  bool insert(Args&&... args) {
+    return emplace_impl(std::forward<Args>(args)...);
+  }
+
+  template <typename... Args>
+  bool emplace(Args&&... args) {
+    return emplace_impl(std::forward<Args>(args)...);
+  }
+
+  // TODO: assert key and value types
+  void merge_other_hashtables(std::vector<std::unique_ptr<LPHashTable>> otherTables) {
+    int total_size = 0;
+    for (const auto& table_ptr : otherTables) {
+      total_size += table_ptr->size();
+    }
+    rehash(total_size);
+    for (const auto& table_ptr : otherTables) {
+      for (auto it = table_ptr->begin(); it != table_ptr->end(); ++it)
+        insert((*it).first.key, (*it).second);
+    }
+  }
+
+  void swap(LPHashTable& other) noexcept {
+    std::swap(buckets_, other.buckets_);
+    std::swap(size_, other.size_);
+    std::swap(empty_key_, other.empty_key_);
+  }
+
+  // Lookup
+  std::vector<mapped_type> lookup(const Key& key) { return lookup_impl(key); }
+
+  // Bucket interface
+  size_type bucket_count() const noexcept { return buckets_.size(); }
+
+  size_type max_bucket_count() const noexcept { return buckets_.max_size(); }
+
+  // Hash policy
+  void rehash(size_type count) {
```

Comment:
when should I call this?

Reply:
Now this function is called in ```emplace_impl```. In theory, rash need to be called when you think the load factor is high. and collision affect the speed